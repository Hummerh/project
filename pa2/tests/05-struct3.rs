struct Point { x: i32, y: i32, z: i32 }

struct Dog {
    loc : Point,
    tailless: bool
}

fn new_dog() -> Dog { Dog { loc: Point {x: 1, y: 2, z: 3}, tailless: true }}

fn main() {
    let p : Point = Point { x: 1, y: 2, z: 3};
    let d : Dog = Dog { loc: Point {x: p.x, y: p.y, z: p.z}, tailless: false };

    d.loc.x * d.loc.y * d.loc.z;
    new_dog().loc.x - new_dog().loc.y + new_dog().loc.z;
    d.tailless || new_dog().tailless;
}
