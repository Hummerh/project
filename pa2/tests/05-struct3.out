(crate
    (items
        (struct-def
            (id
                (Point))
            (field-defs
                (field-def
                    (id
                        (x))
                    (type-i32))
                (field-def
                    (id
                        (y))
                    (type-i32))
                (field-def
                    (id
                        (z))
                    (type-i32))))
        (struct-def
            (id
                (Dog))
            (field-defs
                (field-def
                    (id
                        (loc))
                    (id
                        (Point)))
                (field-def
                    (id
                        (tailless))
                    (type-bool))))
        (fn-def
            (id
                (new_dog))
            (id
                (Dog))
            (block
                (struct
                    (id
                        (Dog))
                    (field-inits
                        (field-init
                            (id
                                (loc))
                            (struct
                                (id
                                    (Point))
                                (field-inits
                                    (field-init
                                        (id
                                            (x))
                                        (lit-dec))
                                    (field-init
                                        (id
                                            (y))
                                        (lit-dec))
                                    (field-init
                                        (id
                                            (z))
                                        (lit-dec)))))
                        (field-init
                            (id
                                (tailless))
                            (true))))))
        (fn-def
            (id
                (main))
            (block
                (let
                    (pat-id
                        (id
                            (p)))
                    (id
                        (Point))
                    (struct
                        (id
                            (Point))
                        (field-inits
                            (field-init
                                (id
                                    (x))
                                (lit-dec))
                            (field-init
                                (id
                                    (y))
                                (lit-dec))
                            (field-init
                                (id
                                    (z))
                                (lit-dec)))))
                (let
                    (pat-id
                        (id
                            (d)))
                    (id
                        (Dog))
                    (struct
                        (id
                            (Dog))
                        (field-inits
                            (field-init
                                (id
                                    (loc))
                                (struct
                                    (id
                                        (Point))
                                    (field-inits
                                        (field-init
                                            (id
                                                (x))
                                            (field-lookup
                                                (id
                                                    (p))
                                                (id
                                                    (x))))
                                        (field-init
                                            (id
                                                (y))
                                            (field-lookup
                                                (id
                                                    (p))
                                                (id
                                                    (y))))
                                        (field-init
                                            (id
                                                (z))
                                            (field-lookup
                                                (id
                                                    (p))
                                                (id
                                                    (z)))))))
                            (field-init
                                (id
                                    (tailless))
                                (false)))))
                (mul
                    (mul
                        (field-lookup
                            (field-lookup
                                (id
                                    (d))
                                (id
                                    (loc)))
                            (id
                                (x)))
                        (field-lookup
                            (field-lookup
                                (id
                                    (d))
                                (id
                                    (loc)))
                            (id
                                (y))))
                    (field-lookup
                        (field-lookup
                            (id
                                (d))
                            (id
                                (loc)))
                        (id
                            (z))))
                (add
                    (sub
                        (field-lookup
                            (field-lookup
                                (fn-call
                                    (id
                                        (new_dog))
                                    ())
                                (id
                                    (loc)))
                            (id
                                (x)))
                        (field-lookup
                            (field-lookup
                                (fn-call
                                    (id
                                        (new_dog))
                                    ())
                                (id
                                    (loc)))
                            (id
                                (y))))
                    (field-lookup
                        (field-lookup
                            (fn-call
                                (id
                                    (new_dog))
                                ())
                            (id
                                (loc)))
                        (id
                            (z))))
                (or
                    (field-lookup
                        (id
                            (d))
                        (id
                            (tailless)))
                    (field-lookup
                        (fn-call
                            (id
                                (new_dog))
                            ())
                        (id
                            (tailless))))))))